include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
+            m_defaultUploadBuffer.get()->multi_place(std::chrono::high_resolution_clock::now() + std::chrono::microseconds(500u), 1u, (const void* const*)&dataPtr, &localOffset, &subSize, &alignment);
+
+            // keep trying again
+            if (localOffset == video::StreamingTransientDataBufferMT<>::invalid_address)
+            {
+                // but first sumbit the already buffered up copies
+                cmdbuf->end();
+                IGPUQueue::SSubmitInfo submit;
+                submit.commandBufferCount = 1u;
+                submit.commandBuffers = &cmdbuf;
+                submit.signalSemaphoreCount = 0u;
+                submit.pSignalSemaphores = nullptr;
+                submit.waitSemaphoreCount = 0u;
+                submit.pWaitSemaphores = nullptr;
+                submit.pWaitDstStageMask = nullptr;
+                queue->submit(1u, &submit, fence); // <threadsafeSubmit==true> means submit while locking an internal mutex
+                waitForFences(1u, &fence, false, 9999999999ull);
+                //event.manuallySignal(); // ???????? just to be exact, probably would still work without, simply because we're resetting a fence that was previously queued up
+                resetFences(1u, &fence);
+                cmdbuf->reset(IGPUCommandBuffer::ERF_RELEASE_RESOURCES_BIT);
+                cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
+                continue;
+            }
+            // some platforms expose non-coherent host-visible GPU memory, so writes need to be flushed explicitly
+            if (m_defaultUploadBuffer.get()->needsManualFlushOrInvalidate())
+            {
+                IDriverMemoryAllocation::MappedMemoryRange flushRange(m_defaultUploadBuffer.get()->getBuffer()->getBoundMemory(), localOffset, subSize);
+                flushMappedMemoryRanges(1u, &flushRange);
+            }
+            // after we make sure writes are in GPU memory (visible to GPU) and not still in a cache, we can copy using the GPU to device-only memory
+            asset::SBufferCopy copy;
+            copy.srcOffset = localOffset;
+            copy.dstOffset = bufferRange.offset + uploadedSize;
+            copy.size = subSize;
+            cmdbuf->copyBuffer(m_defaultUploadBuffer.get()->getBuffer(), bufferRange.buffer.get(), 1u, &copy);
+            // this doesn't actually free the memory, the memory is queued up to be freed only after the GPU fence/event is signalled
+            /*event = */m_defaultUploadBuffer.get()->multi_free(1u, &localOffset, &subSize, core::smart_refctd_ptr<IGPUFence>(fence)); // can queue with a reset but not yet pending fence, just fine
+            uploadedSize += subSize;
+        }
+        //return event.deferred_function; // ????? wtf is deferred_function? why would i have an event here??
```

Comment:
you need a reference/iterator to the event so you can force it to the "done/ready" state for when:
- you plan to reset the fence
- you want to attach a smartpointer to a commandbuffer you do not want to drop until the copy is done